In [1]:
cd ..

e:\GitHub\computervision


In [2]:
import pandas as pd
import numpy as np
import os

In [3]:
datasets_dir = 'E:/Datasets/cor-splits'

In [4]:
df_even = pd.read_csv(os.path.join(datasets_dir, 'even-classes-200/images.csv'))
print(df_even.shape)
df_even.head(3)

(9498, 16)


Unnamed: 0.2  Unnamed: 0.1  Unnamed: 0                  id_video   code  \
0             0         16604      163681  650e811b9ba69cc1a1416de5  278.0   
1             1         32426      233970  650e811b9ba69cc1a141ee8e  326.0   
2             2         31393      225172  650e811b9ba69cc1a141d92e  326.0   

                                              folder  \
0  polygons/flood-probability/2/278/CODE278 2023-...   
1  waze/flood/fec779f2-fc3d-4c41-8809-644258be97b...   
2  waze/flood/47520343-0684-4d06-88d5-dd23a093e59...   

                           file_name  \
0  CODE278 2023-08-27 22-55-21-6.jpg   
1  CODE326 2023-08-20 12-56-32-9.jpg   
2  CODE326 2023-07-29 08-00-12-6.jpg   

                                           file_path  frame_index  \
0  polygons/flood-probability/2/278/CODE278 2023-...           35   
1  waze/flood/fec779f2-fc3d-4c41-8809-644258be97b...           18   
2  waze/flood/47520343-0684-4d06-88d5-dd23a093e59...           17   

               timestamp    initial_timestamp  seen      tags     tag  flood  \
0  2023-08-27 22:55:21.6  2023-08-27 22:55:10  True  ['poça']    poça      0   
1  2023-08-20 12:56:32.9  2023-08-20 12:56:27  True        []  normal      0   
2  2023-07-29 08:00:12.6  2023-07-29 08:00:07  True        []  normal      0   

     set  
0  train  
1  train  
2  train

In [5]:
df_2k = pd.read_csv(os.path.join(datasets_dir, 'sgkf-8-1-1-2000/images.csv'))
print(df_2k.shape)
df_2k.head(3)

(4228, 14)


Unnamed: 0                  id_video    code  \
0       82555  650e811b9ba69cc1a1415c21  1462.0   
1       70375  650e811b9ba69cc1a1415c26  1462.0   
2       82577  650e811b9ba69cc1a1415c21  1462.0   

                                              folder  \
0  comando/bolsao/94390/1462/CODE1462 2023-04-08 ...   
1  comando/bolsao/94390/1462/CODE1462 2023-04-08 ...   
2  comando/bolsao/94390/1462/CODE1462 2023-04-08 ...   

                            file_name  \
0  CODE1462 2023-04-08 11-47-06-3.jpg   
1  CODE1462 2023-04-08 12-11-42-9.jpg   
2  CODE1462 2023-04-08 11-47-13-6.jpg   

                                           file_path  frame_index  \
0  comando/bolsao/94390/1462/CODE1462 2023-04-08 ...           22   
1  comando/bolsao/94390/1462/CODE1462 2023-04-08 ...           21   
2  comando/bolsao/94390/1462/CODE1462 2023-04-08 ...           44   

               timestamp    initial_timestamp  seen tags     tag  flood    set  
0  2023-04-08 11:47:06.3  2023-04-08 11:46:59  True   []  normal      0  train  
1  2023-04-08 12:11:42.9  2023-04-08 12:11:36  True   []  normal      0  train  
2  2023-04-08 11:47:13.6  2023-04-08 11:46:59  True   []  normal      0  train

In [6]:
df_4k = pd.read_csv(os.path.join(datasets_dir, 'sgkf-8-1-1-4000/images.csv'))
print(df_4k.shape)
df_4k.head(3)

(8162, 14)


Unnamed: 0                  id_video    code  \
0      112589  650e811b9ba69cc1a1416011  1462.0   
1       74112  650e811b9ba69cc1a1415c2c  1462.0   
2       82494  650e811b9ba69cc1a1415c1b  1462.0   

                                              folder  \
0  comando/bolsao/97061/1462/CODE1462 2023-05-29 ...   
1  comando/bolsao/94390/1462/CODE1462 2023-04-08 ...   
2  comando/bolsao/94390/1462/CODE1462 2023-04-08 ...   

                            file_name  \
0  CODE1462 2023-05-29 05-56-07-9.jpg   
1  CODE1462 2023-04-08 12-41-08-3.jpg   
2  CODE1462 2023-04-08 11-17-11-9.jpg   

                                           file_path  frame_index  \
0  comando/bolsao/97061/1462/CODE1462 2023-05-29 ...           42   
1  comando/bolsao/94390/1462/CODE1462 2023-04-08 ...           25   
2  comando/bolsao/94390/1462/CODE1462 2023-04-08 ...            6   

               timestamp    initial_timestamp  seen      tags     tag  flood  \
0  2023-05-29 05:56:07.9  2023-05-29 05:55:54  True  ['poça']    poça      0   
1  2023-04-08 12:41:08.3  2023-04-08 12:41:00  True        []  normal      0   
2  2023-04-08 11:17:11.9  2023-04-08 11:17:10  True        []  normal      0   

     set  
0  train  
1  train  
2  train

In [7]:
df_even = df_even[df_even.set == 'train']
df_2k = df_2k[df_2k.set == 'train']
df_4k = df_4k[df_4k.set == 'train']

In [8]:
# df_common_codes = pd.merge(df_even, df_2k, on=['file_path'], how='inner')
# df_common_codes = pd.merge(df_common_codes, df_4k, on=['file_path'], how='inner')

In [9]:
# used_codes = list(set(df_even.code.unique() + df_2k.code.unique() + df_4k.code.unique())) # Doesnt work, the lists need to have same length

used_codes = set()
temp_list = []
for code in [df_even.code.unique(), df_2k.code.unique(), df_4k.code.unique()]:
    temp_list.extend(val for val in code if val not in used_codes)
    used_codes.update(code)
    
print(len(used_codes))

175


In [10]:
df_images = pd.read_csv('data/images.csv')
print(df_images.shape)
df_images.head(3)

(234551, 10)


id_video    code  \
0  65040934f049f672e58adc3b  1461.0   
1  65040934f049f672e58adc3b  1461.0   
2  65040934f049f672e58adc3b  1461.0   

                                              folder  \
0  comando/alagamento/101579/1461/CODE1461 2023-0...   
1  comando/alagamento/101579/1461/CODE1461 2023-0...   
2  comando/alagamento/101579/1461/CODE1461 2023-0...   

                            file_name  \
0  CODE1461 2023-08-28 23-30-02-0.jpg   
1  CODE1461 2023-08-28 23-30-02-3.jpg   
2  CODE1461 2023-08-28 23-30-02-6.jpg   

                                           file_path  frame_index  \
0  comando/alagamento/101579/1461/CODE1461 2023-0...            0   
1  comando/alagamento/101579/1461/CODE1461 2023-0...            1   
2  comando/alagamento/101579/1461/CODE1461 2023-0...            2   

               timestamp    initial_timestamp  seen      tags  
0  2023-08-28 23:30:02.0  2023-08-28 23:30:02  True  ['poça']  
1  2023-08-28 23:30:02.3  2023-08-28 23:30:02  True  ['poça']  
2  2023-08-28 23:30:02.6  2023-08-28 23:30:02  True  ['poça']

In [11]:
df_images_codes = df_images.code.unique()

test_codes = [code for code in df_images_codes if code not in used_codes]
print(len(test_codes))

19


In [68]:
df_test = df_images[df_images['code'].isin(test_codes)]
print(df_test.shape)
df_test.head()

(17031, 10)


id_video   code  \
19819  65040934f049f672e58adf5f  174.0   
19820  65040934f049f672e58adf5f  174.0   
19821  65040934f049f672e58adf5f  174.0   
19822  65040934f049f672e58adf5f  174.0   
19823  65040934f049f672e58adf5f  174.0   

                                                  folder  \
19819  comando/bolsao/97059/174/CODE174 2023-05-29 05...   
19820  comando/bolsao/97059/174/CODE174 2023-05-29 05...   
19821  comando/bolsao/97059/174/CODE174 2023-05-29 05...   
19822  comando/bolsao/97059/174/CODE174 2023-05-29 05...   
19823  comando/bolsao/97059/174/CODE174 2023-05-29 05...   

                               file_name  \
19819  CODE174 2023-05-29 05-10-54-0.jpg   
19820  CODE174 2023-05-29 05-10-54-3.jpg   
19821  CODE174 2023-05-29 05-10-54-6.jpg   
19822  CODE174 2023-05-29 05-10-54-9.jpg   
19823  CODE174 2023-05-29 05-10-55-3.jpg   

                                               file_path  frame_index  \
19819  comando/bolsao/97059/174/CODE174 2023-05-29 05...            0   
19820  comando/bolsao/97059/174/CODE174 2023-05-29 05...            1   
19821  comando/bolsao/97059/174/CODE174 2023-05-29 05...            2   
19822  comando/bolsao/97059/174/CODE174 2023-05-29 05...            3   
19823  comando/bolsao/97059/174/CODE174 2023-05-29 05...            4   

                   timestamp    initial_timestamp  seen tags  
19819  2023-05-29 05:10:54.0  2023-05-29 05:10:54  True   []  
19820  2023-05-29 05:10:54.3  2023-05-29 05:10:54  True   []  
19821  2023-05-29 05:10:54.6  2023-05-29 05:10:54  True   []  
19822  2023-05-29 05:10:54.9  2023-05-29 05:10:54  True   []  
19823  2023-05-29 05:10:55.3  2023-05-29 05:10:54  True   []

In [69]:
df_test['flood'] = df_test['tags'].apply(lambda x: 1 if len(x) > 2 else 0)
print(df_test.shape)
df_test.head()

(17031, 11)


C:\Users\Adm\AppData\Local\Temp\ipykernel_13292\2366281439.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_test['flood'] = df_test['tags'].apply(lambda x: 1 if len(x) > 2 else 0)


id_video   code  \
19819  65040934f049f672e58adf5f  174.0   
19820  65040934f049f672e58adf5f  174.0   
19821  65040934f049f672e58adf5f  174.0   
19822  65040934f049f672e58adf5f  174.0   
19823  65040934f049f672e58adf5f  174.0   

                                                  folder  \
19819  comando/bolsao/97059/174/CODE174 2023-05-29 05...   
19820  comando/bolsao/97059/174/CODE174 2023-05-29 05...   
19821  comando/bolsao/97059/174/CODE174 2023-05-29 05...   
19822  comando/bolsao/97059/174/CODE174 2023-05-29 05...   
19823  comando/bolsao/97059/174/CODE174 2023-05-29 05...   

                               file_name  \
19819  CODE174 2023-05-29 05-10-54-0.jpg   
19820  CODE174 2023-05-29 05-10-54-3.jpg   
19821  CODE174 2023-05-29 05-10-54-6.jpg   
19822  CODE174 2023-05-29 05-10-54-9.jpg   
19823  CODE174 2023-05-29 05-10-55-3.jpg   

                                               file_path  frame_index  \
19819  comando/bolsao/97059/174/CODE174 2023-05-29 05...            0   
19820  comando/bolsao/97059/174/CODE174 2023-05-29 05...            1   
19821  comando/bolsao/97059/174/CODE174 2023-05-29 05...            2   
19822  comando/bolsao/97059/174/CODE174 2023-05-29 05...            3   
19823  comando/bolsao/97059/174/CODE174 2023-05-29 05...            4   

                   timestamp    initial_timestamp  seen tags  flood  
19819  2023-05-29 05:10:54.0  2023-05-29 05:10:54  True   []      0  
19820  2023-05-29 05:10:54.3  2023-05-29 05:10:54  True   []      0  
19821  2023-05-29 05:10:54.6  2023-05-29 05:10:54  True   []      0  
19822  2023-05-29 05:10:54.9  2023-05-29 05:10:54  True   []      0  
19823  2023-05-29 05:10:55.3  2023-05-29 05:10:54  True   []      0

In [70]:
df_test = df_test.sample(frac=1, replace=False, random_state=42)
print(df_test.shape)
df_test.head()

(17031, 11)


id_video    code  \
181047  650e811b9ba69cc1a1418ace  3261.0   
224419  650e811b9ba69cc1a141dacd  3261.0   
181180  650e811b9ba69cc1a1418ade  3261.0   
190150  650e811b9ba69cc1a14196a1   262.0   
101610  650e811b9ba69cc1a1415df9   770.0   

                                                   folder  \
181047  polygons/flood-unlabeled/43/3261/CODE3261 2023...   
224419  waze/flood/56a9d2ab-eec0-4554-9949-9b4556ddbc3...   
181180  polygons/flood-unlabeled/43/3261/CODE3261 2023...   
190150  polygons/flood/19/262/CODE262 2023-02-14 11-30-00   
101610  comando/bolsao/94417/770/CODE770 2023-04-08 10...   

                                 file_name  \
181047  CODE3261 2023-02-07 19-54-02-3.jpg   
224419  CODE3261 2023-07-29 00-50-36-6.jpg   
181180  CODE3261 2023-02-07 22-15-09-9.jpg   
190150   CODE262 2023-02-14 11-30-02-6.jpg   
101610   CODE770 2023-04-08 10-02-52-3.jpg   

                                                file_path  frame_index  \
181047  polygons/flood-unlabeled/43/3261/CODE3261 2023...            7   
224419  waze/flood/56a9d2ab-eec0-4554-9949-9b4556ddbc3...           14   
181180  polygons/flood-unlabeled/43/3261/CODE3261 2023...           30   
190150  polygons/flood/19/262/CODE262 2023-02-14 11-30...            8   
101610  comando/bolsao/94417/770/CODE770 2023-04-08 10...           16   

                    timestamp    initial_timestamp  seen tags  flood  
181047  2023-02-07 19:54:02.3  2023-02-07 19:54:00  True   []      0  
224419  2023-07-29 00:50:36.6  2023-07-29 00:50:32  True   []      0  
181180  2023-02-07 22:15:09.9  2023-02-07 22:15:00  True   []      0  
190150  2023-02-14 11:30:02.6  2023-02-14 11:30:00  True   []      0  
101610  2023-04-08 10:02:52.3  2023-04-08 10:02:47  True   []      0

In [104]:
df_test['file_path'].iloc[0]

'polygons/flood-unlabeled/43/3261/CODE3261 2023-02-07 19-54-00\\CODE3261 2023-02-07 19-54-02-3.jpg'

In [113]:
type(df_test[df_test.file_path == 'polygons/flood-unlabeled/43/3261/CODE3261 2023-02-07 19-54-00\\CODE3261 2023-02-07 19-54-02-3.jpg']['flood'].values[0])

numpy.int64

In [71]:
count_by_code = df_test.groupby('code')['flood'].value_counts().unstack(fill_value=0)
count_by_code

flood      0    1
code             
174.0    457    0
262.0    422   12
414.0      0  225
770.0   1582    0
923.0     60    0
1092.0   322    0
1162.0   139   64
1169.0    17    0
2268.0   101    0
3260.0   405    0
3261.0  8720    0
3262.0   426    0
3263.0   957    0
3264.0    93    0
3265.0  2304    0
3266.0    45    0
3268.0   140    0
3269.0    45    0
3323.0     0  495

The cameras not used in any training are extremaly disbalanced. I will pick at most 100 images for camera for flood = 1 (cameras 262, 414, 1162 and 3323), totalling 281 images. Then, I will choose the same number os images for class 0 between all cameras not used in training.

547 or 557 removed class 1 images

In [90]:
sorted_test = df_test.sort_values(by='code', ascending=False)
sorted_test.shape

(17031, 11)

In [91]:
class_1_codes = [262.0, 414.0, 1162.0, 3323.0]

indexes_to_drop = []
for cls_code in class_1_codes:
    indexes = sorted_test[(sorted_test.code == cls_code) & (sorted_test.flood == 1)].index.tolist()
    if len(indexes) > 100:
        indexes_to_drop += indexes[100:]
sorted_test = sorted_test.drop(indexes_to_drop)
sorted_test.shape

(16511, 11)

In [92]:
count_by_code = sorted_test.groupby('code')['flood'].value_counts().unstack(fill_value=0)
print(count_by_code.shape)
count_by_code

(19, 2)


flood      0    1
code             
174.0    457    0
262.0    422   12
414.0      0  100
770.0   1582    0
923.0     60    0
1092.0   322    0
1162.0   139   64
1169.0    17    0
2268.0   101    0
3260.0   405    0
3261.0  8720    0
3262.0   426    0
3263.0   957    0
3264.0    93    0
3265.0  2304    0
3266.0    45    0
3268.0   140    0
3269.0    45    0
3323.0     0  100

In [93]:
cls_1_total = count_by_code[1].sum()
cls_1_total

276

In [95]:
img_per_camera = int(cls_1_total/(count_by_code.shape[0] - count_by_code[count_by_code[0] == 0].shape[0]))
img_per_camera

16

In [98]:
indexes_to_drop = []
cameras_with_cls_0 = sorted_test[sorted_test.flood == 0]['code'].unique()

for cls_code in cameras_with_cls_0:
    indexes = sorted_test[(sorted_test.code == cls_code) & (sorted_test.flood == 0)].index.tolist()
    indexes_to_drop += indexes[img_per_camera:]
sorted_test = sorted_test.drop(indexes_to_drop)
sorted_test.shape

(548, 11)

In [99]:
count_by_code = sorted_test.groupby('code')['flood'].value_counts().unstack(fill_value=0)
count_by_code

flood    0    1
code           
174.0   16    0
262.0   16   12
414.0    0  100
770.0   16    0
923.0   16    0
1092.0  16    0
1162.0  16   64
1169.0  16    0
2268.0  16    0
3260.0  16    0
3261.0  16    0
3262.0  16    0
3263.0  16    0
3264.0  16    0
3265.0  16    0
3266.0  16    0
3268.0  16    0
3269.0  16    0
3323.0   0  100

In [100]:
sorted_test.to_csv('data/temp_test.csv')

In [102]:
temp_list = sorted_test.file_path.to_list()
print(temp_list)

['polygons/comando/bolsao/30/3323/CODE3323 2023-08-20 13-55-09\\CODE3323 2023-08-20 13-55-11-6.jpg', 'polygons/comando/bolsao/30/3323/CODE3323 2023-08-20 14-16-18\\CODE3323 2023-08-20 14-16-21-3.jpg', 'polygons/comando/bolsao/30/3323/CODE3323 2023-08-20 14-00-27\\CODE3323 2023-08-20 14-00-32-9.jpg', 'polygons/comando/bolsao/30/3323/CODE3323 2023-08-20 13-37-39\\CODE3323 2023-08-20 13-37-49-9.jpg', 'polygons/comando/bolsao/30/3323/CODE3323 2023-08-20 13-52-32\\CODE3323 2023-08-20 13-52-35-9.jpg', 'polygons/comando/bolsao/30/3323/CODE3323 2023-08-20 14-10-11\\CODE3323 2023-08-20 14-10-16-9.jpg', 'polygons/comando/bolsao/30/3323/CODE3323 2023-08-20 13-40-10\\CODE3323 2023-08-20 13-40-10-3.jpg', 'polygons/comando/bolsao/30/3323/CODE3323 2023-08-20 14-00-27\\CODE3323 2023-08-20 14-00-27-0.jpg', 'polygons/comando/bolsao/30/3323/CODE3323 2023-08-20 13-45-14\\CODE3323 2023-08-20 13-45-15-6.jpg', 'polygons/comando/bolsao/30/3323/CODE3323 2023-08-20 13-55-09\\CODE3323 2023-08-20 13-55-19-6.jpg',

In [114]:
import shutil

def copy_images_from_csv(csv_file, output_folder):
    
    base_path = 'E:\GitHub\octa-flood-images\data\images'
    df = pd.read_csv(csv_file)
    
    if not os.path.exists(output_folder): 
        os.makedirs(output_folder)
    
    cls_0_path = os.path.join(output_folder,'0')
    cls_1_path = os.path.join(output_folder,'1')
    
    if not os.path.exists(cls_0_path):
        os.makedirs(cls_0_path)
        
    if not os.path.exists(cls_1_path):
        os.makedirs(cls_1_path)   
    
    for filepath in df.file_path.to_list():
        original_filepath = os.path.join(base_path,filepath)          
        output_filepath = os.path.join(output_folder, str(int(df[df.file_path == filepath]['flood'].values[0])))
        shutil.copy(original_filepath, output_filepath)

In [115]:
cf = 'data/temp_test.csv'
od = 'E:/Datasets/cor-splits/temp-test/test'
copy_images_from_csv(cf, od)